File: sdks/python/example/validation_wizard.py
#!/usr/bin/env python3
"""
DecentriLicense Python SDK 验证向导
================================

这是一个交互式的验证工具，用于测试DecentriLicense Python SDK的所有功能。
"""

import json
import os
import sys
import time
from datetime import datetime
from decenlicense import DecentriLicenseClient, LicenseError


g_client = None
g_initialized = False


class AccountingInfo:
    """记账信息结构"""
    def __init__(self, token_id, device_id, license_code):
        self.token_id = token_id
        self.activation_time = datetime.now().isoformat()
        self.device_id = device_id
        self.license_code = license_code
        self.is_valid = True


class ValidationResult:
    """验证结果结构"""
    def __init__(self, valid=True, token_id=None, error=None, app_id=None, 
                 license_code=None, issue_time=None, expire_time=None, holder_device_id=None):
        self.valid = valid
        self.token_id = token_id
        self.error = error
        self.app_id = app_id
        self.license_code = license_code
        self.issue_time = issue_time
        self.expire_time = expire_time
        self.holder_device_id = holder_device_id


def list_files_for_selection(exts=None):
    files = []
    for name in os.listdir('.'):
        if not os.path.isfile(name):
            continue
        if exts:
            lower = name.lower()
            if not any(lower.endswith(ext) for ext in exts):
                continue
        files.append(name)
    return sorted(files)


def pick_file_from_cwd(prompt_title, exts=None):
    files = list_files_for_selection(exts)
    print(prompt_title)
    if not files:
        return input("当前目录没有可选文件，请手动输入路径: ").strip()

    for i, f in enumerate(files, 1):
        print(f"{i}. {f}")
    print("0. 手动输入路径")

    sel = input("请选择文件编号: ").strip()
    try:
        n = int(sel)
    except ValueError:
        n = -1

    if 1 <= n <= len(files):
        return files[n - 1]
    return input("请输入文件路径: ").strip()


def main():
    """主函数"""
    print("==========================================")
    print("DecentriLicense Python SDK 验证向导")
    print("==========================================")
    print()

    while True:
        print("请选择要执行的操作:")
        print("1. 导入 token(密文/JSON)")
        print("2. 离线验证")
        print("3. 激活绑定本机")
        print("4. 查询状态")
        print("5. 记录使用量/状态迁移")
        print("0. 退出")
        choice = input("请输入选项 (0-5): ").strip()

        if choice == "1":
            activate_token_wizard()
        elif choice == "2":
            verify_token_wizard()
        elif choice == "3":
            accounting_wizard()
        elif choice == "4":
            trust_chain_validation_wizard()
        elif choice == "5":
            comprehensive_validation_wizard()
        elif choice == "0":
            shutdown_client()
            print("感谢使用 DecentriLicense Python SDK 验证向导!")
            break
        else:
            print("无效选项，请重新输入。")

        print("\n" + "-" * 50 + "\n")


def activate_token_wizard():
    """导入 token(密文/JSON)"""
    global g_client, g_initialized
    print("\n--- 1) 导入 token(密文/JSON) ---")

    product_key_file = pick_file_from_cwd("请选择产品公钥文件(含 ROOT_SIGNATURE):", exts=['.pem'])
    if not product_key_file:
        print("产品公钥文件路径不能为空")
        return

    token_file_path = pick_file_from_cwd("请选择 token 文件:", exts=['.txt', '.json'])
    if not token_file_path:
        print("token 文件路径不能为空")
        return

    try:
        with open(product_key_file, 'r', encoding='utf-8') as f:
            product_public_key = f.read()
        with open(token_file_path, 'r', encoding='utf-8') as f:
            token_input = f.read()

        shutdown_client()
        g_client = DecentriLicenseClient()
        g_client.initialize("", udp_port=0, tcp_port=0, registry_server_url="")
        g_client.set_product_public_key(product_public_key)
        g_client.import_token(token_input)
        g_initialized = True

        print("导入成功。产品公钥预览: " + product_public_key[:120] + '...')
        print("token 预览: " + token_input[:120] + '...')
    except Exception as e:
        print(f"导入失败: {e}")
        shutdown_client()


def verify_token_wizard():
    """离线验证"""
    print("\n--- 2) 离线验证 ---")

    if not ensure_initialized():
        return

    try:
        result = g_client.offline_verify_current_token()
        print("离线验证结果: %s" % ("✓ 成功" if result.get('valid') else "✗ 失败"))
        if not result.get('valid'):
            print("错误信息: %s" % result.get('error_message'))
    except Exception as e:
        print(f"离线验证失败: {e}")


def accounting_wizard():
    """激活绑定本机"""
    print("\n--- 3) 激活绑定本机 ---")

    if not ensure_initialized():
        return

    try:
        result = g_client.activate_bind_device()
        print("激活绑定结果: %s" % ("✓ 成功" if result.get('valid') else "✗ 失败"))
        if not result.get('valid'):
            print("错误信息: %s" % result.get('error_message'))
    except Exception as e:
        print(f"激活绑定失败: {e}")


def trust_chain_validation_wizard():
    """查询状态"""
    print("\n--- 4) 查询状态 ---")

    if not ensure_initialized():
        return

    try:
        status = g_client.get_status()
        print("has_token: %s" % int(bool(status.get('has_token'))))
        print("is_activated: %s" % int(bool(status.get('is_activated'))))
        print("state_index: %s" % status.get('state_index'))
        print("token_id: %s" % status.get('token_id'))
        print("holder_device_id: %s" % status.get('holder_device_id'))
        print("app_id: %s" % status.get('app_id'))
        print("license_code: %s" % status.get('license_code'))
    except Exception as e:
        print(f"查询状态失败: {e}")


def comprehensive_validation_wizard():
    """记录使用量/状态迁移"""
    print("\n--- 5) 记录使用量/状态迁移 ---")

    if not ensure_initialized():
        return

    payload_path_or_input = pick_file_from_cwd("请选择 state payload JSON 文件(或 0 手动输入 JSON):", exts=['.json'])
    if not payload_path_or_input:
        print("payload 不能为空")
        return

    payload = payload_path_or_input
    if os.path.isfile(payload_path_or_input):
        try:
            with open(payload_path_or_input, 'r', encoding='utf-8') as f:
                payload = f.read()
        except Exception as e:
            print(f"读取 payload 文件失败: {e}")
            return

    try:
        result = g_client.record_usage(payload)
        print("记录使用量结果: %s" % ("✓ 成功" if result.get('valid') else "✗ 失败"))
        if not result.get('valid'):
            print("错误信息: %s" % result.get('error_message'))
    except Exception as e:
        print(f"记录使用量失败: {e}")


def ensure_initialized():
    if not g_initialized or g_client is None:
        print("请先执行 1) 导入 token(密文/JSON)")
        return False
    return True


def shutdown_client():
    global g_client, g_initialized
    if g_client is not None:
        try:
            g_client.shutdown()
        except Exception:
            pass
    g_client = None
    g_initialized = False


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\n程序被用户中断")
        sys.exit(0)
    except Exception as e:
        print(f"\n程序发生未预期的错误: {e}")
        sys.exit(1)

File: sdks/nodejs/example/validation_wizard.js
#!/usr/bin/env node

/**
 * DecentriLicense Node.js SDK 验证向导
 * 
 * 这是一个交互式的验证工具，用于测试DecentriLicense Node.js SDK的所有功能。
 */

const fs = require('fs');
const path = require('path');
const readline = require('readline');

// 创建 readline 接口
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// 记账信息结构
class AccountingInfo {
    constructor(tokenId, deviceId, licenseCode) {
        this.token_id = tokenId;
        this.activation_time = new Date().toISOString();
        this.device_id = deviceId;
        this.license_code = licenseCode;
        this.is_valid = true;
    }
}

// 验证结果结构
class ValidationResult {
    constructor(valid = true, tokenId = null, error = null, appId = null,
                licenseCode = null, issueTime = null, expireTime = null, holderDeviceId = null) {
        this.valid = valid;
        this.token_id = tokenId;
        this.error = error;
        this.app_id = appId;
        this.license_code = licenseCode;
        this.issue_time = issueTime;
        this.expire_time = expireTime;
        this.holder_device_id = holderDeviceId;
    }
}

// 异步输入函数
function question(prompt) {
    return new Promise((resolve) => {
        rl.question(prompt, resolve);
    });
}

function listFilesForSelection(cwd) {
    const entries = fs.readdirSync(cwd, { withFileTypes: true });
    return entries
        .filter((e) => e.isFile())
        .map((e) => e.name)
        .sort((a, b) => a.localeCompare(b));
}

async function pickFileFromCwd(title, exts) {
    const cwd = process.cwd();
    const files = listFilesForSelection(cwd).filter((f) => {
        if (!exts || exts.length === 0) return true;
        const lower = f.toLowerCase();
        return exts.some((ext) => lower.endsWith(ext));
    });

    console.log(title);
    if (files.length === 0) {
        const manual = await question("当前目录没有可选文件，请手动输入路径: ");
        return (manual || '').trim();
    }

    for (let i = 0; i < files.length; i++) {
        console.log(`${i + 1}. ${files[i]}`);
    }
    console.log("0. 手动输入路径");

    let sel = await question("请选择文件编号: ");
    sel = (sel || '').trim();
    const n = parseInt(sel, 10);
    if (!isNaN(n) && n >= 1 && n <= files.length) {
        return path.join(cwd, files[n - 1]);
    }
    const manual = await question("请输入文件路径: ");
    return (manual || '').trim();
}

async function main() {
    console.log("==========================================");
    console.log("DecentriLicense Node.js SDK 验证向导");
    console.log("==========================================");
    console.log();

    while (true) {
        console.log("请选择要执行的操作:");
        console.log("1. 激活令牌");
        console.log("2. 校验令牌");
        console.log("3. 记账信息");
        console.log("4. 信任链验证");
        console.log("5. 综合验证");
        console.log("6. 退出");

        try {
            const choice = await question("请输入选项 (1-6): ");

            switch (choice.trim()) {
                case "1":
                    await activateTokenWizard();
                    break;
                case "2":
                    await verifyTokenWizard();
                    break;
                case "3":
                    await accountingWizard();
                    break;
                case "4":
                    await trustChainValidationWizard();
                    break;
                case "5":
                    await comprehensiveValidationWizard();
                    break;
                case "6":
                    console.log("感谢使用 DecentriLicense Node.js SDK 验证向导!");
                    rl.close();
                    return;
                default:
                    console.log("无效选项，请重新输入。");
            }
        } catch (error) {
            console.error("操作失败:", error.message);
        }

        console.log("\n" + "-".repeat(50) + "\n");
    }
}

async function activateTokenWizard() {
    console.log("\n--- 激活令牌 ---");
    
    // 获取许可证代码
    let licenseCode = await question("请输入许可证代码 (默认: EXAMPLE-LICENSE-12345): ");
    licenseCode = licenseCode.trim() || "EXAMPLE-LICENSE-12345";

    // 获取UDP端口
    let udpPortStr = await question("请输入UDP端口 (默认: 8888): ");
    let udpPort = 8888;
    if (udpPortStr.trim() && !isNaN(parseInt(udpPortStr.trim()))) {
        udpPort = parseInt(udpPortStr.trim());
    }

    // 获取TCP端口
    let tcpPortStr = await question("请输入TCP端口 (默认: 8889): ");
    let tcpPort = 8889;
    if (tcpPortStr.trim() && !isNaN(parseInt(tcpPortStr.trim()))) {
        tcpPort = parseInt(tcpPortStr.trim());
    }

    try {
        // 动态导入SDK
        const { default: DecentriLicenseClient } = await import('./index.js');
        const client = new DecentriLicenseClient();
        
        console.log("初始化客户端...");
        client.initialize({
            licenseCode: licenseCode,
            udpPort: udpPort,
            tcpPort: tcpPort,
            registryServerUrl: ""
        });
        
        const deviceId = client.getDeviceId();
        console.log(`设备ID: ${deviceId}`);

        // 激活许可证
        console.log("激活许可证...");
        console.log("广播局域网发现...");
        console.log("等待对等节点和选举...");

        try {
            const result = client.activate();
            const success = result && typeof result === 'object' ? result.success : Boolean(result);
            console.log(`激活成功: ${success}`);
            
            // 记录记账信息
            if (success) {
                const accounting = new AccountingInfo(
                    "generated-token-id", // 实际应用中应从激活结果获取
                    deviceId,
                    licenseCode
                );
                
                // 保存记账信息到文件
                const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
                const filename = `accounting_${timestamp}.json`;
                fs.writeFileSync(filename, JSON.stringify(accounting, null, 2), 'utf8');
                console.log(`记账信息已保存到: ${filename}`);
            }
        } catch (error) {
            console.log(`激活失败: ${error.message}`);
        }
        
        // 清理资源
        client.shutdown();
        
    } catch (error) {
        console.log(`操作失败: ${error.message}`);
    }
}

async function verifyTokenWizard() {
    console.log("\n--- 校验令牌 ---");
    
    // 获取令牌文件路径
    const tokenFilePath = await pickFileFromCwd("请选择 token 文件:", ['.txt', '.json']);
    
    if (!tokenFilePath.trim()) {
        console.log("令牌文件路径不能为空");
        return;
    }

    try {
        // 读取令牌文件
        if (!fs.existsSync(tokenFilePath.trim())) {
            console.log("找不到指定的令牌文件");
            return;
        }
        
        const tokenData = fs.readFileSync(tokenFilePath.trim(), 'utf8');

        // 这里应该调用SDK的令牌验证功能
        // 由于当前SDK实现简化，我们模拟验证过程
        const preview = tokenData.length > 100 ? tokenData.substring(0, 100) + "..." : tokenData;
        console.log(`令牌内容预览: ${preview}`);
        
        // 模拟验证结果
        const result = new ValidationResult(
            true,
            "sample-token-id",
            null,
            "sample-app-id",
            "EXAMPLE-LICENSE-12345",
            new Date(Date.now() - 24 * 60 * 60 * 1000).toLocaleString(),
            new Date(Date.now() + 365 * 24 * 60 * 60 * 1000).toLocaleString(),
            "sample-device-id"
        );

        // 显示验证结果
        console.log("\n验证结果:");
        if (result.valid) {
            console.log("✓ 令牌验证成功");
            console.log(`  令牌ID: ${result.token_id}`);
            console.log(`  应用ID: ${result.app_id}`);
            console.log(`  许可证代码: ${result.license_code}`);
            console.log(`  签发时间: ${result.issue_time}`);
            console.log(`  过期时间: ${result.expire_time}`);
            console.log(`  持有设备ID: ${result.holder_device_id}`);
        } else {
            console.log("✗ 令牌验证失败");
            console.log(`  错误信息: ${result.error}`);
        }
    } catch (error) {
        console.log(`校验令牌时发生错误: ${error.message}`);
    }
}

async function accountingWizard() {
    console.log("\n--- 记账信息 ---");
    
    // 查找记账文件
    const files = fs.readdirSync('.');
    const accountingFiles = files.filter(file => 
        file.startsWith('accounting_') && file.endsWith('.json')
    );

    if (accountingFiles.length === 0) {
        console.log("未找到记账文件");
        return;
    }

    console.log("找到以下记账文件:");
    accountingFiles.forEach((file, index) => {
        console.log(`${index + 1}. ${file}`);
    });

    try {
        const choiceStr = await question("请选择要查看的文件编号: ");
        const choice = parseInt(choiceStr.trim());
        
        if (choice >= 1 && choice <= accountingFiles.length) {
            const selectedFile = accountingFiles[choice - 1];
            
            const data = fs.readFileSync(selectedFile, 'utf8');
            const accounting = JSON.parse(data);
            
            console.log("\n记账信息详情:");
            console.log(`令牌ID: ${accounting.token_id || 'N/A'}`);
            console.log(`激活时间: ${accounting.activation_time || 'N/A'}`);
            console.log(`设备ID: ${accounting.device_id || 'N/A'}`);
            console.log(`许可证代码: ${accounting.license_code || 'N/A'}`);
            console.log(`是否有效: ${accounting.is_valid || 'N/A'}`);
        } else {
            console.log("无效选择");
        }
    } catch (error) {
        console.log(`读取记账信息时发生错误: ${error.message}`);
    }
}

async function trustChainValidationWizard() {
    console.log("\n--- 信任链验证 ---");
    console.log("信任链验证用于验证令牌的完整信任链...");
    
    // 模拟信任链验证过程
    console.log("1. 验证根公钥签名...");
    console.log("2. 验证产品公钥签名...");
    console.log("3. 验证令牌签名...");
    console.log("4. 验证环境一致性...");
    
    // 模拟验证结果
    console.log("\n信任链验证结果:");
    console.log("✓ 根公钥签名验证通过");
    console.log("✓ 产品公钥签名验证通过");
    console.log("✓ 令牌签名验证通过");
    console.log("✓ 环境一致性验证通过");
    console.log("✓ 信任链完整且有效");
}

async function comprehensiveValidationWizard() {
    console.log("\n--- 综合验证 ---");
    console.log("综合验证将执行完整的许可证验证流程...");
    
    // 模拟综合验证过程
    console.log("1. 检查系统环境...");
    console.log("2. 验证网络连接...");
    console.log("3. 加载许可证...");
    console.log("4. 验证信任链...");
    console.log("5. 检查设备状态...");
    console.log("6. 验证许可证有效性...");
    
    // 模拟验证结果
    console.log("\n综合验证结果:");
    console.log("✓ 系统环境检查通过");
    console.log("✓ 网络连接正常");
    console.log("✓ 许可证加载成功");
    console.log("✓ 信任链验证通过");
    console.log("✓ 设备状态正常");
    console.log("✓ 许可证有效");
    console.log("✓ 综合验证完成");
}

// 运行主程序
if (require.main === module) {
    main().catch(console.error);
}

module.exports = { main };